producer/buffer/buffer.go

```diff
@@ -0,0 +1,197 @@
+		bytesBuffered:   scope.Gauge("bytes-buffered"),
+	}
+}
+
+type buffer struct {
+	sync.Mutex
+
+	strategy        OnFullStrategy
+	maxBufferSize   uint64
+	cleanupInterval time.Duration
+	buffer          *list.List
+	m               *bufferMetrics
+
+	size     *atomic.Uint64
+	doneCh   chan struct{}
+	isClosed bool
+}
+
+// NewBuffer returns a new buffer.
+func NewBuffer(opts Options) producer.Buffer {
+	if opts == nil {
+		opts = NewBufferOptions()
+	}
+	return &buffer{
+		size:            atomic.NewUint64(0),
+		buffer:          list.New(),
+		isClosed:        false,
+		doneCh:          make(chan struct{}),
+		strategy:        opts.OnFullStrategy(),
+		maxBufferSize:   opts.MaxBufferSize(),
+		cleanupInterval: opts.CleanupInterval(),
+		m:               newBufferMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+}
+
+func (b *buffer) Buffer(d producer.Data) (producer.RefCountedData, error) {
+	b.Lock()
+	if b.isClosed {
+		b.Unlock()
+		return nil, errBufferClosed
+	}
+	var (
+		dataSize      = uint64(d.Size())
+		maxBufferSize = b.maxBufferSize
+	)
+	if dataSize > maxBufferSize {
+		b.Unlock()
+		return nil, errInvalidDataLargerThanMaxBufferSize
+	}
+	targetBufferSize := maxBufferSize - dataSize
+	if b.size.Load() > targetBufferSize {
+		if err := b.produceOnFullWithLock(targetBufferSize); err != nil {
+			b.Unlock()
+			return nil, err
+		}
+	}
+	b.size.Add(dataSize)
+	rd := data.NewRefCountedData(d, b.subSize)
+	b.buffer.PushBack(rd)
+	b.Unlock()
+	return rd, nil
+}
+
+func (b *buffer) produceOnFullWithLock(targetSize uint64) error {
+	switch b.strategy {
+	case ReturnError:
+		return errBufferFull
+	case DropEarliest:
+		b.dropEarliestUntilTargetWithLock(targetSize)
+	}
+	return nil
+}
+
+func (b *buffer) dropEarliestUntilTargetWithLock(targetSize uint64) {
+	var next *list.Element
+	for e := b.buffer.Front(); e != nil && b.size.Load() > targetSize; e = next {
+		next = e.Next()
+		d := e.Value.(producer.RefCountedData)
+		if !d.IsClosed() {
+			b.m.messageDropped.Inc(1)
+			b.m.bytesDropped.Inc(int64(d.Size()))
+			d.Drop()
+		}
+		b.buffer.Remove(e)
+	}
+}
+
+func (b *buffer) Init() {
+	go b.cleanupForever()
+}
+
+func (b *buffer) cleanupForever() {
+	ticker := time.NewTicker(b.cleanupInterval)
+	for {
+		select {
+		case <-ticker.C:
+			b.Lock()
+			b.cleanupWithLock()
+			b.m.messageBuffered.Update(float64(b.buffer.Len()))
+			b.Unlock()
+			b.m.bytesBuffered.Update(float64(b.size.Load()))
+		case <-b.doneCh:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+func (b *buffer) cleanupWithLock() {
+	var next *list.Element
+	for e := b.buffer.Front(); e != nil; e = next {
+		next = e.Next()
+		d := e.Value.(producer.RefCountedData)
+		if d.IsClosed() {
+			b.buffer.Remove(e)
+		}
+	}
+}
+
+func (b *buffer) Close() {
+	// Stop taking writes right away.
+	b.Lock()
+	if b.isClosed {
+		b.Unlock()
+		return
+	}
+	b.isClosed = true
+	b.Unlock()
+
+	// Block until all data consumed.
+	for {
+		b.Lock()
```

Comment:
Hm this seems CPU intensive. Do you not need some sort of backoff? Otherwise it'll be spinning locking.

Reply:
Would it be possible to use a notification channel? Delay helps but still costs CPU.

Reply:
Yeah sounds good